### PR TITLE
health check: gracefully handle GOAWAY in grpc health checker

### PR DESCRIPTION
**Pull request by jmuia**
Commit Message:
Gracefully handle GOAWAY in gRPC health checker, allowing in-progress healthchecks to complete before closing the connection.

Additional Description:
We've observed failures in our production and integration testing environments due to the gRPC health checker interpreting GOAWAYs as a health check failure. As an example, if all endpoints in an upstream cluster drain their incoming connections at the same time (eg. listener config change) a grpc healthcheck client with `unhealthy_threshold=1` will see the entire cluster become unhealthy, which is undesirable.

This is my first contribution to Envoy, and would appreciate a close eye for edge cases I may have missed!

Risk Level: low/medium
Testing: added unit tests, ran (failing) grpc hc internal integration test with this patch which passed
Docs Changes: n/a
Release Notes: n/a

Signed-off-by: Joey Muia <joseph.b.muia@gmail.com>

### CODE REVIEW

**Comment by murray-stripe** [comment]
@lizan These tests seem to be a flake. Or at least I can't seem to reproduce the failure locally with either
```
bazel test //test/common/config:subscription_factory_impl_test
# or
./ci/do_ci.sh 'bazel.release' //test/common/config:subscription_factory_impl_test
```
Would you mind taking a look, or kicking off the azure-pipeline tests again?

**Comment by murray-stripe** [comment]
@lizan Similar to #11324 - I could not reproduce the test-failures using the following commands:
```shell
./ci/do_ci.sh 'bazel.release' //test/common/config:subscription_factory_impl_test
# or
bazel test //test/common/config:subscription_factory_impl_test
```

The test-failure is the same as #11324 and doesn't seem related to the change. I'm not sure if this is a common flake or if I am missing something with my test commands. If it's likely flakey, could you kick off the azure pipeline run again?

File: source/common/upstream/health_checker_impl.cc [hunk 1]
```diff
@@ -725,13 +732,12 @@ void GrpcHealthCheckerImpl::GrpcActiveHealthCheckSession::onResetStream(Http::St
 void GrpcHealthCheckerImpl::GrpcActiveHealthCheckSession::onGoAway() {
   ENVOY_CONN_LOG(debug, "connection going away health_flags={}", *client_,
                  HostUtility::healthFlagsToString(*host_));
-  // Even if we have active health check probe, fail it on GOAWAY and schedule new one.
+  // If we have an active health check probe, allow it to complete before closing the connection.
   if (request_encoder_) {
-    handleFailure(envoy::data::core::v3::NETWORK);
-    expect_reset_ = true;
-    request_encoder_->getStream().resetStream(Http::StreamResetReason::LocalReset);
+    received_goaway_ = true;
```
Comment (snowp) [inline source/common/upstream/health_checker_impl.cc:737]: A one liner comment here about when this connection would eventually get closed would be nice. IIUC it's either on the next timeout or when the remote closes it?

File: source/common/upstream/health_checker_impl.cc [hunk 2]
```diff
@@ -757,6 +763,8 @@ void GrpcHealthCheckerImpl::GrpcActiveHealthCheckSession::onRpcComplete(
     handleFailure(envoy::data::core::v3::ACTIVE);
   }
 
+  const bool goaway = received_goaway_;
```
Comment (snowp) [inline source/common/upstream/health_checker_impl.cc:766]: Mind adding a comment here explaining why we have to read the value here? Its a bit less clear than on L705 since we're not unconditionally calling `resetState`

File: source/common/upstream/health_checker_impl.cc [hunk 3]
```diff
@@ -777,13 +785,18 @@ void GrpcHealthCheckerImpl::GrpcActiveHealthCheckSession::resetState() {
   request_encoder_ = nullptr;
   decoder_ = Grpc::Decoder();
   health_check_response_.reset();
+  received_goaway_ = false;
 }
 
 void GrpcHealthCheckerImpl::GrpcActiveHealthCheckSession::onTimeout() {
   ENVOY_CONN_LOG(debug, "connection/stream timeout health_flags={}", *client_,
                  HostUtility::healthFlagsToString(*host_));
   expect_reset_ = true;
```
Comment (snowp) [inline source/common/upstream/health_checker_impl.cc:794]: should this be inside the `else` branch?

File: test/common/upstream/health_checker_impl_test.cc [hunk 4]
```diff
@@ -4529,21 +4529,176 @@ TEST_F(GrpcHealthCheckerImplTest, GrpcFailUnknownHealthStatus) {
             cluster_->prioritySet().getMockHostSet(0)->hosts_[0]->health());
 }
 
-// Test receiving GOAWAY is interpreted as connection close event.
+// Test receiving GOAWAY is handled gracefully while a check is in progress.
```
Comment (snowp) [inline test/common/upstream/health_checker_impl_test.cc:4532]: This tells me that the code intentionally treated GOAWAY as failures, which makes me wonder if this is the desired behavior for some? Does this mean that this should be configurable?

@baranov1ch @htuch @mattklein123 WDYT

Comment (htuch) [reply test/common/upstream/health_checker_impl_test.cc:4532]: This might depend on the reason for the GOAWAY? I.e. a `NO_ERROR` mid check might be fine as it's telling to drain, but not if it was an error.

File: source/common/upstream/health_checker_impl.cc (hunk 1 above)
Comment (jmuia) [reply source/common/upstream/health_checker_impl.cc:737]: I'll add a comment. It would be closed when the active check completes (`onRpcComplete`) or other conditions like `onResetStream` and `onTimeout`.

File: source/common/upstream/health_checker_impl.cc (hunk 3 above)
Comment (jmuia) [reply source/common/upstream/health_checker_impl.cc:794]: I don't believe so. It's my understanding that calling `client_->close();` here will fire the `onResetStream` callback (and so we want it to return early).

File: test/common/upstream/health_checker_impl_test.cc (hunk 4 above)
Comment (jmuia) [reply test/common/upstream/health_checker_impl_test.cc:4532]: I see, that makes sense. 

It looks like the GOAWAY error code isn't provided in the callback at the moment, but I think it would be reasonable to do that?

It seems that the `onGoAway()` callbacks are only fired once (see [here](https://github.com/envoyproxy/envoy/blob/f71ab4cfb1a7293f2ffb3a7271550d719ee01393/source/common/http/http2/codec_impl.cc#L580-L587)), and - with error codes - we'd want to call them for each GOAWAY frame received. I'm not sure if that would be problematic. For reference it looks the "fire once" behaviour was added in https://github.com/envoyproxy/envoy/pull/103.

It may also be worth mentioning that the http2 connection pool implementation doesn't change behaviour based on the GOAWAY error code at the moment (see [here](https://github.com/envoyproxy/envoy/blob/f71ab4cfb1a7293f2ffb3a7271550d719ee01393/source/common/http/http2/conn_pool.cc#L27-L37)). It transitions the active client to the draining state if it has active requests.

Comment (mattklein123) [reply test/common/upstream/health_checker_impl_test.cc:4532]: Right, if the GOAWAY error code is NO_ERROR we should just treat it as "don't reuse connection", not an error. We may have to add some additional error code info to the callback. Note that @antoniovicente just opened https://github.com/envoyproxy/envoy/issues/11420, though IIRC nghttp2 does provide the reason code inside the frame here: https://github.com/envoyproxy/envoy/blob/5d8f7d785b59d1450b67f3279706b787c5641a66/source/common/http/http2/codec_impl.cc#L582, we just need to store it.

Comment (jmuia) [reply test/common/upstream/health_checker_impl_test.cc:4532]: > IRC nghttp2 does provide the reason code inside the frame here:

Yep - it looks like it's provided in `frame->goaway.error_code` and the debug data in `frame->goaway.opaque_data`


I'm happy to update the `onGoAway()` interface as part of this PR if @antoniovicente doesn't already have something started.

Does something along these lines look right? 
```
  # Maybe pass an Envoy-internal enum rather than the nghttp2 enum?
  virtual void onGoAway(nghttp2_error_code error_code, absl::Span<uint8_t> debug_data) PURE;
```

Comment (mattklein123) [reply test/common/upstream/health_checker_impl_test.cc:4532]: We should not leak nghttp2 specific structs to the public interface, so at least for the error code we should do a mapping. For your use case maybe just have an enum with {NO_ERROR, OTHER} and @antoniovicente can fill out more stuff later as needed?

Comment (antoniovicente) [reply test/common/upstream/health_checker_impl_test.cc:4532]: enum with {NO_ERROR, OTHER} sounds like a good start

File: source/common/http/http2/codec_impl.cc [hunk 5]
```diff
@@ -577,12 +587,10 @@ int ConnectionImpl::onFrameReceived(const nghttp2_frame* frame) {
     }
   }
 
-  // Only raise GOAWAY once, since we don't currently expose stream information. Shutdown
-  // notifications are the same as a normal GOAWAY.
-  if (frame->hd.type == NGHTTP2_GOAWAY && !raised_goaway_) {
+  // Shutdown notifications are the same as a normal GOAWAY.
+  if (frame->hd.type == NGHTTP2_GOAWAY) {
```
Comment (jmuia) [inline source/common/http/http2/codec_impl.cc:591]: I'm not sure if this is fine or if we want to track which goaway error codes we've already seen, and run the callbacks once per error code.

File: test/common/http/http2/codec_impl_test.cc [hunk 6]
```diff
@@ -1456,7 +1457,7 @@ TEST_P(Http2CodecImplTest, LargeRequestHeadersExceedPerHeaderLimit) {
   request_headers.addCopy("big", long_string);
 
   EXPECT_CALL(request_decoder_, decodeHeaders_(_, _)).Times(0);
-  EXPECT_CALL(client_callbacks_, onGoAway());
+  EXPECT_CALL(client_callbacks_, onGoAway(_));
```
Comment (jmuia) [inline test/common/http/http2/codec_impl_test.cc:1460]: I don't completely follow why this test doesn't trigger `onGoAway(_)` to be called twice like the other one.

File: include/envoy/http/codec.h [hunk 7]
```diff
@@ -36,6 +36,11 @@ const char MaxResponseHeadersCountOverrideKey[] =
 
 class Stream;
 
+enum class ErrorCode {
```
[the remark this answers is not in the document]
Comment (jmuia) [reply include/envoy/http/codec.h:39]: 👍 will update. Initially this enumerated all the error codes [here](https://http2.github.io/http2-spec/#ErrorCodes), which are shared with RST_STREAM. But we've since undone that and there's already a separate Envoy-internal enum for stream resets.

File: source/common/http/http2/codec_impl.cc (hunk 5 above)
Comment (jmuia) [reply source/common/http/http2/codec_impl.cc:591]: It is possible to receive GOAWAY frames with different error codes:

> An endpoint MAY send multiple GOAWAY frames if circumstances change. For instance, an endpoint that sends GOAWAY with NO_ERROR during graceful shutdown could subsequently encounter a condition that requires immediate termination of the connection.

(from https://http2.github.io/http2-spec/#GOAWAY)

File: include/envoy/http/codec.h [hunk 8]
```diff
@@ -36,6 +36,11 @@ const char MaxResponseHeadersCountOverrideKey[] =
 
 class Stream;
 
+enum class GoAwayErrorCode {
```
Comment (snowp) [inline include/envoy/http/codec.h:39]: sorry should have mention this but this should have a doc comment as its in `include/`

File: source/common/http/http2/codec_impl.cc (hunk 5 above)
Comment (snowp) [reply source/common/http/http2/codec_impl.cc:591]: Hmm in that case should the logic for failing the HC be that we see a NO_ERROR + no error GOAWAYs?

Also I'm not sure what the implications are of making `onGoAway` trigger multiple times per stream, maybe @mattklein123 or @alyssawilk has an idea? I'd be worried about invariants elsewhere relying on this being called at most once.

Comment (jmuia) [reply source/common/http/http2/codec_impl.cc:591]: > Hmm in that case should the logic for failing the HC be that we see a NO_ERROR + no error GOAWAYs?

Hm, I don't quite follow, could you rephrase?

> I'd be worried about invariants elsewhere relying on this being called at most once.

Agreed. It seems it was added in https://github.com/envoyproxy/envoy/pull/103. I'm not sure if that was an optimization back then. I'm also not sure if it has become an expected behaviour (intentionally or not) since then.

Comment (snowp) [reply source/common/http/http2/codec_impl.cc:591]:
```
  if (request_encoder_ && error_code == Http::GoAwayErrorCode::NoError) {
    received_no_error_goaway_ = true;
    return;
  }
```

if a GOAWAY error follows a NO_ERROR then `received_no_error_goaway_` would still be true. I was thinking that the logic should reset this flag if it saw an error frame.

Comment (jmuia) [reply source/common/http/http2/codec_impl.cc:591]: Ah, I see. I don't expect that to change the behaviour as it stands:
* If there isn't an active probe, the first (NO_ERROR) GOAWAY would've caused the connection to be closed (and we only set `received_no_error_goaway_ = true;` when there's an active probe). 
* If there is an active probe, the stream reset will end up resetting state (`resetState()` will be called by `onResetStream()`).
* In general, I believe a connection close will end up resetting state because the `onResetStream()` callbacks will run ([from here](https://github.com/envoyproxy/envoy/blob/0125d6fd611958fdef85e6665f307d2bffd2a531/source/common/http/codec_client.cc#L71)).

I admit this isn't obvious / does feel a bit brittle -- happy to add a comment and/or reset the flag anyways.

**Comment by jmuia** [comment]
@snowp the CI failures look like flakes. Can you re-run them?

File: source/common/http/http2/codec_impl.cc (hunk 5 above)
Comment (jmuia) [reply source/common/http/http2/codec_impl.cc:591]: Friendly ping @mattklein123 @alyssawilk

Comment (mattklein123) [reply source/common/http/http2/codec_impl.cc:591]: Sorry for the delay. My preference would be to not raise multiple goaways from the codec as I'm guessing there will be code that won't expect this and will break.

Can we just keep the previous behavior and not handle the case where multiple go away frames are received? We can potentially deal with this in a follow up if someone requests it?

Comment (jmuia) [reply source/common/http/http2/codec_impl.cc]: Thanks for the feedback - that seems reasonable to me. I've reverted that specific commit and added a TODO instead.

**Comment by jmuia** [comment]
@snowp ready for another review. I reverted the commit that allowed multiple GOAWAYs and added a TODO instead.
